### ISSUE TITLE
How to exclude files with a certain name? (implement recursive globs)

### ISSUE THREAD

**Issue by ZsoltMolnarrr** [realm/SwiftLint#2329]
I want to exclude from linting every file named "R.generated.swift", but I can't figure out the proper exclude entry for this. Pleas help me!:D

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? - SwiftLint (0.25.1) (from Podfile.lock)
* Installation method used (Homebrew, CocoaPods, building from source, etc)? - CocoaPods
* Paste your configuration file:

```yml
excluded: # paths to ignore during linting. Takes precedence over `included`.
    - Pods
    - MyFeatureKit/iOS/Source/Resources/R.generated.swift # works but very inconvenient
    - MyApp/Source/Resources/R.generated.swift # works but very inconvenient
    # - */R.generated.swift # Doesn't work
    # - */*/R.generated.swift # Doesn't work
    # - .*/R.generated.swift # Doesn't work
```
* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents. - **NO** (My repository only has one .swiftlint.yml at the root of the repository.)
* Which Xcode version are you using (check `xcode-select -p`)? - Xcode 9.4.1

**Comment by marcelofabri** [comment]
This is possible on SwiftLint 0.26. I'm closing this for now, feel free to reopen if you update and can't make it work.

**Comment by ZsoltMolnarrr** [comment]
@marcelofabri thanks for the reply!
Same configuration, but updated to SwiftLint 0.27.
Tried the following entries:
`    - */R.generated.swift"` -> Causes error: Scanner: while scanning an alias in line #, column 7
`    - **/R.generated.swift"` -> Causes error: Scanner: while scanning an alias in line #, column 7
`    - */*/R.generated.swift"` -> Causes error: Scanner: while scanning an alias in line #, column 7
`    - R.generated.swift` -> Generated file is not getting ignored
`    - /*R.generated.swift` -> Generated file is not getting ignored
`    - .*/R.generated.swift` -> Generated file is not getting ignored
`    - ./*/R.generated.swift` -> Generated file is not getting ignored

`- ./*/*/*/*/R.generated.swift` - Works but this requires knowledge about the file hierarchy! Source of different build targets may contain the generated source files at a different hierarchy level.

I except this should work with the same logic as the following .gitignore entry:
`R.generated.swift`
The gitingore logic of the line below would be even better.
`*.generated.swift`

Can you please give me some hints how exactly use it?

**Comment by marcelofabri** [comment]
@ZsoltMolnarrr the errors you're getting is because you need to escape the string on yaml:

```yml
excluded:
  - "*.generated.swift"
```

However, I think what you want is not currently supported because it'd need recursive globs. Let's keep this open as an enhancement.

**Comment by GregHilston** [comment]
@marcelofabri You're solution above is not ignoring my `R.generated.swift` file. Would this be related to my `.swiftlint.yml` being in a child folder, relative to `R.generated.swift`?

**Comment by bdrelling** [comment]
Any chance of bumping this back up? I can't do any of the following:

```
excluded: 
  - "**/*/_*.swift"
  - "**/*/*.generated.swift"
  - "_*.swift"
  - "*.generated.swift"
```

My current workaround is just moving them into other folders, but I'd like to keep some generated files alongside their counterparts. Similarly, we use another tool that uses the `_` prefix for generated files, so we want to account for both.

**Comment by bdrelling** [comment]
@marcelofabri Any thoughts on how to work around this?

**Comment by pgorzelany-objectivity** [comment]
Same issue

**Comment by marcelofabri** [comment]
PRs are welcome 😊

**Comment by RolandasRazma** [comment]
> @marcelofabri Any thoughts on how to work around this?

we work rounded by generating list of files ourselves and then passing list to SwiftLint instead of just path to dir to scan

**Comment by kuyazee** [comment]
Any updates for this issue?

**Comment by dantes-git** [comment]
Any update on this issue?

**Comment by jbdtky** [comment]
Any update?

**Comment by nahung89** [comment]
I think this issue is fixed already. This config works on my side:
```
excluded:
  - "**/*/R.generated.swift"
  ```

I'm using SwiftLint 0.42.0. The project I'm working on separate R.generated.swift file in multiple sub-projects

**Comment by rogerluan** [comment]
I'm on 0.43.1 and I don't think this works yet. `**/*/something` not only fails to match all the "*something", but it's also the wrong glob syntax. The right syntax should be "**/something" and this one also doesn't work

**Comment by cameroncooke** [comment]
Why is this marked as "Won't Fix" when clearly many people and expressing the need for this?

**Comment by rogerluan** [comment]
Because it was a bot that closed it and no one came here to help us reopen it 😞

**Comment by niorko** [comment]
still does not work properly

**Comment by mostafaelshazly9** [comment]
I believe I have found a solution, at least one that worked for me:
TLDR:
add
`- builds`
to your list of exclusions ALONG with the path you are already using (example at the very end of the post)
**### Cause of the issue:**
So, 
`excluded:
  - "*.generated.swift"
`
does work, in my case it cut the errors and warnings to about half.
Why half? because it did ignore the generated file, however, there was another generated file being generated at runtime then moved immediately after, (guessing that's just how RSwift does thing).
To get the exact path of the "intermediary generated file" do this:
1- build app
2- click on a warning/error for the R.generated.swift
3- on the top left corner, you will notice "R.generated.swift" without a previous path
4- **right click** on to see where it originated, it will show the path vertically, it's gonna be something like:
R.generated.swift
<Path>
project-<number, probably 0>
<number, probably 0>
builds

The entire "builds" folder is generated and removed on build, but during that time it detects both of them and thus produces the warning

### **to fix this issue, you can add this to your yml file:**
`excluded:
  - "R.generated.swift"
  - builds
`

**Comment by rogerluan** [comment]
Recursive glob still doesn't work. Use case:

Folder structure:

```
.
├── Pods
└── another_directory
    ├── Pods
    └── somefile.txt
```

How can I ignore both `Pods` directories, considering that one is nested, the other is at the top level? In other words, how can I ignore all directories called `Pods` regardless of where they are? 
I've tried these, and none of them work (individually):

```
- Pods // Catches only the directory at the top level
- **/Pods // Doesn't work 
- */**/Pods // Only catches Pods directory nested on the 3rd level
- .Pods // Doesn't work
- /Pods // Doesn't work
- */Pods // Only catches Pods directory nested on the 2nd level
```

Long story short it seems like I have to know beforehand how many levels nested my folders are going to be to add them explicitly, e.g.:

```
- Pods // level 0
- */Pods // level 1
- */**/Pods // level 2
- */**/**/Pods // level 3
- */**/**/**/Pods // level 4
…
```

Whereas with glob, this should work for all levels: `**/Pods`

[This tool helps](https://www.digitalocean.com/community/tools/glob?comments=true&glob=%2A%2A%2FPods&matches=false&tests=%2F%2F%20This%20will%20match%20as%20it%20ends%20with%20%27.js%27&tests=%2Fhello%2Fworld.js&tests=%2F%2F%20This%20won%27t%20match%21&tests=%2Ftest%2Fsome%2Fglobs&tests=%2FPods&tests=Pods&tests=%2FAnotherDir%2FPods&tests=%2FSomething%2FAnotherDir%2FPods&tests=Something%2FAnotherDir%2FPods):

<img width="429" alt="image" src="https://user-images.githubusercontent.com/8419048/142634521-16924242-3040-46a9-8dea-28d1aa5efc53.png">

----

Could someone reopen this issue? 🙏

**Comment by timwredwards** [comment]
Still broken 🙁

**Comment by hlung** [comment]
WHERE IS THE SOLUTION???

**Comment by hlung** [comment]
> This is possible on SwiftLint 0.26. I'm closing this for now, feel free to reopen if you update and can't make it work.

 @marcelofabri Any example?

**Comment by jkmathew** [comment]
This works with `"**/*generated.swift"`.

For the directory tree like this, all `*.generated.swift` has been excluded.

```
── App
│   └── AppTests
│       ├── Mocks
│       │   ├── A.swift
│       │   ├── B.swift
│       │   ├── C.swift
│       │   ├── AAA.generated.swift
│       │   ├── Generated
│       │   │   ├── A.generated.swift
│       │   │   └── B.generated.swift
```

I am using Version `0.49.1`

**Comment by JanC** [comment]
I tried `"**/*generated.swift` as suggested above but it does seem to match my file located at `Projects/MyFeature/Sources/GraphQL/API.generated.swift`

In fact, even if I specify the file explicitly, it is still linted/corrected

```yaml
excluded: 
  - "Projects/MyFeature/Sources/GraphQL/API.generated.swift"
 ```

I'm using version `0.50.3`

Am I doing wrong in my config?

**Comment by Ravichandrane** [comment]
Any update ? 
Still broken on 0.50.3

**Comment by lhunath** [comment]
Version `0.50.3`, this leads to a `Bus error`:
```
Linting Swift files in current working directory
Bus error: 10
```
Config:
```
included:
  - Sources
excluded:
  - "**/*.graphql.swift"
```
LLDB:
```
(lldb) f 5
frame #5: 0x0000000100152e28 swiftlint`static SwiftLintFramework.Glob.expandGlobstar(pattern: Swift.String) -> Swift.Array<Swift.String> + 604
swiftlint`static SwiftLintFramework.Glob.expandGlobstar(pattern: Swift.String) -> Swift.Array<Swift.String>:
->  0x100152e28 <+604>: mov    x21, x0
    0x100152e2c <+608>: mov    x0, x20
    0x100152e30 <+612>: bl     0x10075ac20               ; symbol stub for: objc_release
    0x100152e34 <+616>: mov    x0, x27
(lldb) f 4
frame #4: 0x000000010015381c swiftlint`function signature specialization <Arg[1] = Owned To Guaranteed, Arg[2] = Dead> of function signature specialization <Arg[0] = [Closure Propagated : closure #1 (Swift.String) -> Swift.Optional<Swift.String> in static SwiftLintFramework.Glob.(expandGlobstar in _A29680EDB2A67C45FA4E37C1A6FCC33B)(pattern: Swift.String) -> Swift.Array<Swift.String>, Argument Types : [Swift.String@thin SwiftLintFramework.Glob.Type]> of generic specialization <Swift.Array<Swift.String>, Swift.String> of Swift.Sequence.compactMap<τ_0_0>((τ_0_0.Element) throws -> Swift.Optional<τ_1_0>) throws -> Swift.Array<τ_1_0> + 308
swiftlint`function signature specialization <Arg[1] = Owned To Guaranteed, Arg[2] = Dead> of function signature specialization <Arg[0] = [Closure Propagated : closure #1 (Swift.String) -> Swift.Optional<Swift.String> in static SwiftLintFramework.Glob.(expandGlobstar in _A29680EDB2A67C45FA4E37C1A6FCC33B)(pattern: Swift.String) -> Swift.Array<Swift.String>, Argument Types : [Swift.String@thin SwiftLintFramework.Glob.Type]> of generic specialization <Swift.Array<Swift.String>, Swift.String> of Swift.Sequence.compactMap<τ_0_0>((τ_0_0.Element) throws -> Swift.Optional<τ_1_0>) throws -> Swift.Array<τ_1_0>:
->  0x10015381c <+308>: mov    x22, x0
    0x100153820 <+312>: mov    x0, x19
    0x100153824 <+316>: bl     0x10075ac20               ; symbol stub for: objc_release
    0x100153828 <+320>: mov    x0, x20
(lldb) f 3
frame #3: 0x00000001993535ec Foundation`-[NSFileManager fileExistsAtPath:isDirectory:] + 56
Foundation`-[NSFileManager fileExistsAtPath:isDirectory:]:
->  0x1993535ec <+56>: cbz    w0, 0x19935363c           ; <+136>
    0x1993535f0 <+60>: movi.2d v0, #0000000000000000
    0x1993535f4 <+64>: stp    q0, q0, [sp, #0x70]
    0x1993535f8 <+68>: stp    q0, q0, [sp, #0x50]
(lldb) f 2
frame #2: 0x000000019932e9d4 Foundation`-[NSFileManager getFileSystemRepresentation:maxLength:withPath:] + 52
Foundation`-[NSFileManager getFileSystemRepresentation:maxLength:withPath:]:
->  0x19932e9d4 <+52>: tbz    w0, #0x0, 0x19932e9e0     ; <+64>
    0x19932e9d8 <+56>: mov    w0, #0x0
    0x19932e9dc <+60>: b      0x19932ea04               ; <+100>
    0x19932e9e0 <+64>: mov    x0, x21
(lldb) f 1
frame #1: 0x00000001993300d0 Foundation`-[NSPathStore2 isEqualToString:] + 272
Foundation`-[NSPathStore2 isEqualToString:]:
->  0x1993300d0 <+272>: adrp   x10, 370197
    0x1993300d4 <+276>: add    x10, x10, #0x2f0          ; _MergedGlobals
    0x1993300d8 <+280>: ldrsw  x8, [x10, #0x4]
    0x1993300dc <+284>: ldr    w9, [x19, x8]
(lldb) f 0
frame #0: 0x000000019841325c CoreFoundation`-[__NSCFString getCharacters:range:] + 4
CoreFoundation`-[__NSCFString getCharacters:range:]:
->  0x19841325c <+4>:  stp    x20, x19, [sp, #-0x20]!
    0x198413260 <+8>:  stp    x29, x30, [sp, #0x10]
    0x198413264 <+12>: add    x29, sp, #0x10
    0x198413268 <+16>: mov    x8, x2
```

**Comment by mildm8nnered** [comment]
Hmmm. In the SwiftLint source tree, with these extra files:

So this all appears to work as expected for me with `0.50.3` and also `0.51.0`, on Monterey, with Xcode 14.2

```
% find .   
.
./t.yml
./A
./A/A
./A/A/X.graphql.swiftx
./A/X.graphql.swift
./X.graphql.swift
./B
./B/Foo.swift
% 
% 
% cat t.yml
included:
  - A
  - B  
excluded:
  - "**/*.graphql.swift"
% 
% 
%  /opt/homebrew/bin/swiftlint --config t.yml 
Linting Swift files in current working directory
Linting 'Foo.swift' (1/1)
Done linting! Found 0 violations, 0 serious in 1 file.
```

**Comment by mildm8nnered** [comment]
Similarly with directories

```
% find . 
.
./t.yml
./A
./A/AFoo.swift
./A/B
./A/B/ABFoo.swift
./B
./B/BFoo.swift
% 
% 
% cat t.yml 
included:
  - A
  - B  
excluded:
  - "**/B"
% 
% 
%  /opt/homebrew/bin/swiftlint --config t.yml 
Linting Swift files in current working directory
Linting 'AFoo.swift' (1/1)
Done linting! Found 0 violations, 0 serious in 1 file.
```

**Comment by lazarevzubov** [comment]
I have multiple files called "Something" inside different child folders on different levels of the hierarchy, and this works for me:

```yaml
excluded:
  - '**/something'
```
 
I'm using SwiftLint 0.54.0.

**Comment by rogerluan** [comment]
> I have multiple files called "Something" inside different child folders on different levels of the hierarchy, and this works for me:
> 
> excluded:
>   - '**/something'
> I'm using SwiftLint 0.54.0.

+1!

I'm using this and it works great:

https://github.com/rogerluan/.linters/blob/3b2cc66dded86d23505da434bb7349c059e30e87/.swiftlint.yml#L16-L24

SwiftLint 0.51.0

**Comment by rogerluan** [comment]
With that, I think this issue can be closed 😃!
